Remove field duplicates suggested changes

darrynten opened the pull request:
This is opened into #17 

## Overview

| Q                 | A
| ----------------- | ---
| Bugfix?           | yes
| New feature?      | yes
| Breaking Changes? | no
| Fixed issues      | -

## The problem

Reviewing @make-it-git PR brought some issues to light

* Added additional example fields after studying SageOne docs some more
* Changed `persistable` to `readonly`
* Formatted tests
* Fixed a broken true/false check in the base tests
* Consistent variable interpolation
* Updated assertion strings
* Updated example mock to match new assertions
* Added validation and tests for min/max and regex checks

## Notes

The new rules are:

* `min` and `max` - both must be present, validates an integer is in a range, or that a string length is in a range
* `regex` - ensures a string matches a pattern

NB regarding 'required' this is here for discussion

Please look and let me know if it's all understood please @make-it-git

darrynten committed:
Added github template for Issues and Pull Requests (#18)

Added github template for Issues and Pull Requests

This will help ensure consistent PRs and Issues, and will help keep any
new developers joining the project on the same page.

darrynten committed:
Additional example fields

darrynten committed:
Replaced 'persistable' with 'readonly'

Read only makes more sense than persistable as it is what the
official documentation uses.

Inverted all appropriate booleans.

darrynten committed:
Errant file

darrynten committed:
Formatting of tests

darrynten committed:
Fixed test true/false attribute check

darrynten committed:
Updated BaseModelTest docblock indicating attributes are all mandatory

darrynten committed:
Consistent string/variable interpolation

darrynten committed:
Match assertion strings

darrynten committed:
Updated example mock to match expectations

darrynten commented on src/Models/Example.php:
```diff
@@ -77,13 +70,41 @@ class Example extends BaseModel
+            'type' => 'string',
+            'nullable' => false,
+            'readonly' => false,
+            'required' => true,
```
This is intended as a discussion point - is this really needed?

darrynten commented on src/Models/Example.php:
```diff
@@ -77,13 +70,41 @@ class Example extends BaseModel
+            'min' => 0,
+            'max' => 100,
+            'regex' => "/^[A-Za-z0-9,!#\$%&'\*\+/=\?\^_`\{\|}~-]+(\.[A-Za-z0-9,!#\$%&'\*\+/=\?\^_`\{\|}~-]+)*@[A-Za-z0-9-]+(\.[A-Za-z0-9-]+)*\.([A-Za-z]{2,})$/",
+        ]
```
These need validation added, will happen in different PR

darrynten commented on tests/SageOne/Models/AccountNoteModelTest.php:
```diff
@@ -63,31 +63,49 @@ public function testAttributes()
-                'type' => 'boolean', 'nullable' => true, 'persistable' => true
+                'type' => 'boolean',
+                'nullable' => true,
+                'readonly' => false,
```
The logic was breaking a bit here as it was not explicit like it is now, there were assumed defaults

darrynten replied:
The other new possible attributes should be checked here too

darrynten commented on tests/SageOne/Models/BaseModelTest.php:
```diff
@@ -189,23 +189,32 @@ protected function verifyAttributes(string $class, array $attributes)
+            $this->assertEquals('boolean', gettype($value[$name]['readonly']));
 
-            $nullable = isset($options['nullable']);
+            $nullable = $options['nullable'];
```
This is why we need to be explicit, we can't leave nullable or read-only as optional parameters, as checking if it was set was not working 100% in some cases

darrynten committed:
Added validation and tests for min/max and regex optional field params

darrynten committed:
More defensive test examples

darrynten commented on tests/SageOne/Models/ExampleModelTest.php:
```diff
@@ -357,14 +384,54 @@ public function testExamples()
+        $this->assertArrayHasKey('regex', $value['emailAddress']);
+        $this->assertEquals('string', gettype($value['emailAddress']['regex']));
+        $this->assertRegExp($value['emailAddress']['regex'], 'me@example.com');
+        $this->assertRegExp($value['emailAddress']['regex'], 'another.valid+email@example.com');
```
These are examples of good, defensive checks that should always be performed 🔮

darrynten commented on tests/mocks/Example/GET_Example_Get_xx.json:
```diff
@@ -1,4 +1,10 @@
+  "RequiredString": "required information",
+  "EmailAddress": "user@example.com",
+  "StringWithDefault": "default string changed",
+  "SomeBoolean": true
```
Expand on this example mock instead of using reflection to modify

Feel free to use multiple mocks, replacing the `xx` in the filename with the required id

darrynten commented on src/BaseModel.php:
```diff
@@ -417,6 +490,69 @@ private function isValidPrimitive($resultItem, $definedType)
+        }
+    }
+
+
```
I left too many newline :(

make-it-git commented on src/BaseModel.php:
```diff
@@ -95,29 +95,93 @@ public function __construct(array $config)
-        if (!$this->fields[$key]['nullable'] && is_null($value)) {
-            $this->throwException(ModelException::NULL_WITHOUT_NULLABLE, sprintf('attempting to nullify key %s', $key));
+        if (gettype($value) === 'string') {
+            if ((strlen($value) < $min) || (strlen($value) > $max)) {
```
Should not we use mb_strlen?

darrynten replied:
You're absolutely correct will update

darrynten committed:
Mocks for various HTTP Response Codes

I have added what I can while I await my API Key.

If your API key works feel free to add mocks to this branch.

- Reference https://www.sageone.co.za/api-response-codes

darrynten committed:
Merge branch 'dev' into remove-field-duplicates-darryn

* dev:
  Mocks for various HTTP Response Codes
  Added github template for Issues and Pull Requests (#18)

darrynten committed:
Use multibyte string length checks

darrynten committed:
Errant newline
